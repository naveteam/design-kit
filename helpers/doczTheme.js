import React from 'react'

import { Provider } from 'src/Theme'

export default ({ children }) => {
  return <Provider>{children}</Provider>
}
