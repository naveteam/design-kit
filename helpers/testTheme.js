import React from 'react'
import { render } from '@testing-library/react'

import { Provider } from 'src/Theme'

export function renderWithTheme(component) {
  return render(<Provider>{component}</Provider>)
}
