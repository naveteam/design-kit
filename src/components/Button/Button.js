import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import modifiers from 'src/helpers/modifiers'

const PRIMARY = 'primary'

const SECONDARY = 'secondary'

const GHOST = 'ghost'

const DefaultButton = styled.button`
  border: 0;
  outline: none;
  height: 32px;
  font-weight: 500;
  transition: all 0.3s ease-in-out;
  width: 140px;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  font-family: Roboto;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: normal;
  text-align: center;
`

const Button = props => {
  const { variant, ...rest } = props
  if (variant === PRIMARY) {
    return <PrimaryButton {...rest} />
  }

  if (variant === SECONDARY) {
    return <SecondaryButton {...rest} />
  }

  if (variant === GHOST) {
    return <GhostButton {...rest} />
  }
  return <PrimaryButton {...rest} />
}

const PrimaryButton = styled(DefaultButton)`
  ${props => {
    const { theme } = props
    return `
      ${modifiers(props)}
      background-color: ${theme.colors.blue};
      color: ${theme.colors.white};
      &:hover {
        background-color: ${theme.colors.darkBlue};
      }
      &:focus {
        box-shadow: 0 0 4px 4px rgba(0, 125, 254, 0.3);
      }
      &:disabled {
        background-color: ${theme.colors.grey};
        color: rgba(0, 0, 0, 0.3);
        &:hover {
          cursor: not-allowed;
        }
      }
    `
  }};
`

const SecondaryButton = styled(DefaultButton)`
  ${props => {
    const { theme } = props
    return `
      ${modifiers(props)}
      background-color: ${theme.colors.white};
      color: ${theme.colors.blue};
      border: 1px solid ${theme.colors.blue};
      &:hover {
        color: ${theme.colors.white};
        background-color: ${theme.colors.blue};
      }
      &:focus {
        box-shadow: 0 0 4px 4px rgba(0, 125, 254, 0.3);
      }
      &:disabled {
        background-color: ${theme.colors.white};
        color: rgba(0, 0, 0, 0.3);
        border: 1px solid ${theme.colors.grey};
        &:hover {
          cursor: not-allowed;
        }
      }
    `
  }};
`

const GhostButton = styled(DefaultButton)`
  ${props => {
    const { theme } = props
    return `
      ${modifiers(props)}
      background-color: ${theme.colors.white};
      color: ${theme.colors.blue};
      &:hover {
        background-color: ${theme.colors.grey};
      }
      &:focus {
        box-shadow: 0 0 4px 4px rgba(0, 125, 254, 0.3);
        :hover {
          background-color: ${theme.colors.white};
        }
      }
      &:disabled {
        background-color: ${theme.colors.white};
        color: rgba(0, 0, 0, 0.3);
        &:hover {
          cursor: not-allowed;
        }
      }
    `
  }};
`

Button.propTypes = {
  variant: PropTypes.string
}

export default Button
