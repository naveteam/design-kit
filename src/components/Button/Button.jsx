import React, { forwardRef } from 'react'
import styled, { css } from '@xstyled/styled-components'
import { layout, variant, th } from '@xstyled/system'
import PropTypes from 'prop-types'

import { Typography } from '../'

const Button = forwardRef(({ children, ...props }, ref) => (
  <Base ref={ref} {...props}>
    <Typography fontSize={2} fontWeight={1} lineHeight={3}>
      {children}
    </Typography>
  </Base>
))

const colorVariants = variant({
  default: 'filled',
  key: 'button',
  variants: {
    filled: css`
      background-color: ${({ color }) => th.color(color)};
      border-color: ${({ color }) => th.color(color)};
      color: white;
      &:hover {
        background-color: ${({ color }) => th.color(`${color}_hover`)};
        border-color: ${({ color }) => th.color(`${color}_hover`)};
      }
      &:active {
        background-color: ${({ color }) => th.color(`${color}_active`)};
        border-color: ${({ color }) => th.color(`${color}_active`)};
      }
      &:disabled {
        background-color: disabled;
        border-color: disabled;
        color: white;
      }
    `,
    outlined: css`
      background-color: transparent;
      border-color: ${({ color }) => th.color(color)};
      color: ${({ color }) => th.color(color)};
      &:hover {
        border-color: ${({ color }) => th.color(`${color}_hover`)};
        color: ${({ color }) => th.color(`${color}_hover`)};
      }
      &:active {
        border-color: ${({ color }) => th.color(`${color}_active`)};
        color: ${({ color }) => th.color(`${color}_active`)};
      }
      &:disabled {
        border-color: disabled;
        color: disabled;
      }
    `,
    text: css`
      background-color: transparent;
      border-color: transparent;
      color: ${({ color }) => th.color(color)};
      &:hover {
        color: ${({ color }) => th.color(`${color}_hover`)};
      }
      &:active {
        color: ${({ color }) => th.color(`${color}_active`)};
      }
      &:disabled {
        color: disabled;
      }
    `
  }
})

const Base = styled.button`
  cursor: pointer;
  padding: 2;
  border-width: 1px;
  border-style: solid;
  border-radius: 2;
  min-height: 40px;
  ${layout}
  ${colorVariants}
  &:disabled {
    cursor: initial;
  }
  &:focus {
    outline: none;
  }
`

Button.defaultProps = {
  color: 'primary',
  variant: 'filled',
  width: 1
}

Button.propTypes = {
  color: PropTypes.oneOf(['primary', 'secondary']),
  variant: PropTypes.oneOf(['filled', 'outlined', 'text']),
  width: PropTypes.oneOfType([PropTypes.arrayOf([PropTypes.number]), PropTypes.number])
}

export default Button
