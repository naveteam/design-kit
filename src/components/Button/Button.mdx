---
name: Button
---

import { Playground, Props } from 'docz'
import Button from './Button'

# Button

<Props of={Button} />

## Primary

<Playground>
  <Button variant='primary'>Click me</Button>
</Playground>

## Secondary

<Playground>
  <Button variant='secondary'>Click me</Button>
</Playground>

## Ghost

<Playground>
  <Button variant='ghost'>Click me</Button>
</Playground>
