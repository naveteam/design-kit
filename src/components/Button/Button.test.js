import React from 'react'
import { renderWithTheme } from 'helpers/testTheme'

import Button from './Button'

describe('Button', () => {
  it('renders without pass children', () => {
    const { container } = renderWithTheme(<Button />)
    expect(container).toMatchSnapshot()
  })

  it('renders with primary variant', () => {
    const { container } = renderWithTheme(<Button variant='primary'>test</Button>)
    expect(container).toMatchSnapshot()
  })

  it('renders with secondary variant', () => {
    const { container } = renderWithTheme(<Button variant='secondary'>test</Button>)
    expect(container).toMatchSnapshot()
  })

  it('renders with ghost variant', () => {
    const { container } = renderWithTheme(<Button variant='ghost'>test</Button>)
    expect(container).toMatchSnapshot()
  })

  it('renders disabled', () => {
    const { container } = renderWithTheme(<Button disabled>test</Button>)

    expect(container.firstChild).toHaveAttribute('disabled')
    expect(container).toMatchSnapshot()
  })
})
