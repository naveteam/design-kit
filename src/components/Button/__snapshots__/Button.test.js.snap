// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders disabled 1`] = `
.c0 {
  border: 0;
  outline: none;
  height: 32px;
  font-weight: 500;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  width: 140px;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  font-family: Roboto;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  text-align: center;
  background-color: #007dfe;
  color: #ffffff;
}

.c0:hover {
  background-color: #076cd5;
}

.c0:focus {
  box-shadow: 0 0 4px 4px rgba(0,125,254,0.3);
}

.c0:disabled {
  background-color: #f5f5f5;
  color: rgba(0,0,0,0.3);
}

.c0:disabled:hover {
  cursor: not-allowed;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    test
  </button>
</div>
`;

exports[`Button renders with ghost variant 1`] = `
.c0 {
  border: 0;
  outline: none;
  height: 32px;
  font-weight: 500;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  width: 140px;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  font-family: Roboto;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  text-align: center;
  background-color: #ffffff;
  color: #007dfe;
}

.c0:hover {
  background-color: #f5f5f5;
}

.c0:focus {
  box-shadow: 0 0 4px 4px rgba(0,125,254,0.3);
}

.c0:focus:hover {
  background-color: #ffffff;
}

.c0:disabled {
  background-color: #ffffff;
  color: rgba(0,0,0,0.3);
}

.c0:disabled:hover {
  cursor: not-allowed;
}

<div>
  <button
    class="c0"
  >
    test
  </button>
</div>
`;

exports[`Button renders with primary variant 1`] = `
.c0 {
  border: 0;
  outline: none;
  height: 32px;
  font-weight: 500;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  width: 140px;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  font-family: Roboto;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  text-align: center;
  background-color: #007dfe;
  color: #ffffff;
}

.c0:hover {
  background-color: #076cd5;
}

.c0:focus {
  box-shadow: 0 0 4px 4px rgba(0,125,254,0.3);
}

.c0:disabled {
  background-color: #f5f5f5;
  color: rgba(0,0,0,0.3);
}

.c0:disabled:hover {
  cursor: not-allowed;
}

<div>
  <button
    class="c0"
  >
    test
  </button>
</div>
`;

exports[`Button renders with secondary variant 1`] = `
.c0 {
  border: 0;
  outline: none;
  height: 32px;
  font-weight: 500;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  width: 140px;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  font-family: Roboto;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  text-align: center;
  background-color: #ffffff;
  color: #007dfe;
  border: 1px solid #007dfe;
}

.c0:hover {
  color: #ffffff;
  background-color: #007dfe;
}

.c0:focus {
  box-shadow: 0 0 4px 4px rgba(0,125,254,0.3);
}

.c0:disabled {
  background-color: #ffffff;
  color: rgba(0,0,0,0.3);
  border: 1px solid #f5f5f5;
}

.c0:disabled:hover {
  cursor: not-allowed;
}

<div>
  <button
    class="c0"
  >
    test
  </button>
</div>
`;

exports[`Button renders without pass children 1`] = `
.c0 {
  border: 0;
  outline: none;
  height: 32px;
  font-weight: 500;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  width: 140px;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  font-family: Roboto;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  text-align: center;
  background-color: #007dfe;
  color: #ffffff;
}

.c0:hover {
  background-color: #076cd5;
}

.c0:focus {
  box-shadow: 0 0 4px 4px rgba(0,125,254,0.3);
}

.c0:disabled {
  background-color: #f5f5f5;
  color: rgba(0,0,0,0.3);
}

.c0:disabled:hover {
  cursor: not-allowed;
}

<div>
  <button
    class="c0"
  />
</div>
`;
